src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java

```diff
@@ -307,151 +368,114 @@ final DualNode cmpExNext(DualNode cmp, DualNode val) {
             return (DualNode)NEXT.compareAndExchange(this, cmp, val);
         }
 
+        /** Returns true if this node has been matched or cancelled  */
+        final boolean matched() {
+            return isData != (item != null);
+        }
+
         /**
-         * Returns true if this node has not been matched (or cancelled)
+         * Relaxed write to replace reference to user data with
+         * self-link. Can be used only if not already null after
+         * match.
          */
-        final boolean isLive() {
-            return isData == (item != null);
+        final void selfLinkItem() {
+            ITEM.set(this, this);
         }
 
-        // Relaxed writes when volatile is unnecessarily strong
-        final void forgetItem()        { ITEM.set(this, this); }
-        final void selfLink()          { NEXT.set(this, this); }
-        final void setNext(DualNode n) { NEXT.set(this, n);  }
+        /** The number of times to spin when eligible */
+        private static final int SPINS = 1 << 7;
 
-        // ManagedBlocker support
-        public final boolean isReleasable() {
-            return (!isLive() || Thread.currentThread().isInterrupted());
-        }
-        public final boolean block() {
-            while (!isReleasable()) LockSupport.park();
-            return true;
-        }
+        /**
+         * The number of nanoseconds for which it is faster to spin
+         * rather than to use timed park. A rough estimate suffices.
+         */
+        private static final long SPIN_FOR_TIMEOUT_THRESHOLD = 1L << 10;
+
+        /**
+         * True if system is a uniprocessor, occasionally rechecked.
+         */
+        private static boolean isUniprocessor =
```

Comment:
@DougLea  If this is intentionally non-volatile, I think it is worth documenting the rationale.

Reply:
Added: (L348)
     * than unnecessarily requiring volatile accesses elsewhere. This
     * fence also separates accesses to field isUniprocessor.